envoy/http/header_validator.h

```diff
@@ -74,36 +73,100 @@ class HeaderValidator {
    * The gRPC request will still be rejected with the INTERNAL (13) error code.
    */
-  using RequestHeaderMapValidationResult = RejectOrRedirectResult;
   virtual RequestHeaderMapValidationResult
   validateRequestHeaderMap(RequestHeaderMap& header_map) PURE;
 
   /**
    * Validate the entire response header map.
-   * Returning the Reject value causes the HTTP request to be rejected with the 502 status,
+   * The header map can not be modified in-place as it is immutable after the terminal encoder
+   * filter. However HTTP/2 and HTTP/3 header validators may need to change the status of the
+   * upgrade upstream response to the extended CONNECT downstream response. In this case the new
+   * header map is returned in the `new_headers` member of the returned structure. Returning the
+   * Reject value in the status member causes the HTTP request to be rejected with the 502 status,
    * and the gRPC request with the UNAVAILABLE (14) error code.
    */
-  using ResponseHeaderMapValidationResult = RejectResult;
-  virtual ResponseHeaderMapValidationResult
-  validateResponseHeaderMap(ResponseHeaderMap& header_map) PURE;
+  struct ConstResponseHeaderMapValidationResult {
+    RejectResult status;
+    ResponseHeaderMapPtr new_headers;
+  };
+  virtual ConstResponseHeaderMapValidationResult
+  validateResponseHeaderMap(const ResponseHeaderMap& header_map) PURE;
 
   /**
    * Validate the entire request trailer map.
    * Returning the Reject value causes the HTTP request to be rejected with the 502 status,
    * and the gRPC request with the UNAVAILABLE (14) error code.
    * If response headers have already been sent the request is reset.
    */
-  using TrailerValidationResult = RejectResult;
   virtual TrailerValidationResult validateRequestTrailerMap(RequestTrailerMap& trailer_map) PURE;
 
+  /**
+   * Note: Response trailers are not validated after they were processed by the encoder filter
+   * chain.
+   */
+};
+
+/**
+ * Interface for server header validators.
+ */
+class ClientHeaderValidator : public HeaderValidatorBase {
+public:
+  ~ClientHeaderValidator() override = default;
+
+  /**
+   * Validate the entire request header map.
+   * This method can not mutate the header map as it is immutable after the terminal decoder filter.
+   * However HTTP/2 and HTTP/3 header validators may need to change the request from the HTTP/1
+   * upgrade to to the extended CONNECT. In this case the new header map is returned in the
+   * `new_headers` member of the returned structure. Returning the Reject value form this method
+   * causes the HTTP request to be rejected with 400 status, and the gRPC request with the INTERNAL
+   * (13) error code.
+   */
+  struct ConstRequestHeaderMapValidationResult {
+    RejectResult status;
+    RequestHeaderMapPtr new_headers;
+  };
+  virtual ConstRequestHeaderMapValidationResult
+  validateRequestHeaderMap(const RequestHeaderMap& header_map) PURE;
```

Comment:
stepping back a bit I want to have a high level discussion about validation vs normalization.
It's surely more efficient to do the two together but from an API perspective feels a lot cleaner to do it in 2 steps.
Also don't we do more than just connect based transforms?  I thought the whole drop_headers_with_underscore path would need new headers as well?
/wait-any